### ISSUE TITLE
Installing pyomniscidbe brings two versions of thrift

### ISSUE THREAD

**Issue by vnlitvinov** [conda-forge/omniscidb-feedstock#53]
- [x] I read [the conda-forge documentation](https://conda-forge.org/docs/user/introduction.html#how-can-i-install-packages-from-conda-forge) and could not find the solution for my problem there.

Issue: `conda create -n test pyomniscidbe -c conda-forge` brings two different thrift engines: `libthrift-0.14` and `thrift-cpp-0.13`.
In some environments this could lead to conflicting behaviour.

<br/>
Environment (<code>conda list</code>):
<details>

```
$ conda list
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                       1_gnu    conda-forge
abseil-cpp                20210324.2           h9c3ff4c_0    conda-forge
arrow-cpp                 3.0.0           py39h69b4fad_24_cpu    conda-forge
arrow-cpp-proc            3.0.0                       cpu    conda-forge
aws-c-cal                 0.5.11               h95a6274_0    conda-forge
aws-c-common              0.6.2                h7f98852_0    conda-forge
aws-c-event-stream        0.2.7               h3541f99_13    conda-forge
aws-c-io                  0.10.5               hfb6a706_0    conda-forge
aws-checksums             0.1.11               ha31a3da_7    conda-forge
aws-sdk-cpp               1.8.186              hb4091e7_3    conda-forge
binutils_impl_linux-64    2.36.1               h193b22a_2    conda-forge
binutils_linux-64         2.36                 hf3e587d_1    conda-forge
blosc                     1.21.0               h9c3ff4c_0    conda-forge
boost-cpp                 1.74.0               h312852a_4    conda-forge
bzip2                     1.0.8                h7f98852_4    conda-forge
c-ares                    1.17.2               h7f98852_0    conda-forge
ca-certificates           2021.5.30            ha878542_0    conda-forge
cairo                     1.16.0            h6cf1ce9_1008    conda-forge
cfitsio                   3.470                hb418390_7    conda-forge
curl                      7.79.0               hea6ffbf_0    conda-forge
cyrus-sasl                2.1.27               h230043b_2    conda-forge
double-conversion         3.1.5                h9c3ff4c_2    conda-forge
expat                     2.4.1                h9c3ff4c_0    conda-forge
font-ttf-dejavu-sans-mono 2.37                 hab24e00_0    conda-forge
font-ttf-inconsolata      3.000                h77eed37_0    conda-forge
font-ttf-source-code-pro  2.038                h77eed37_0    conda-forge
font-ttf-ubuntu           0.83                 hab24e00_0    conda-forge
fontconfig                2.13.1            hba837de_1005    conda-forge
fonts-conda-ecosystem     1                             0    conda-forge
fonts-conda-forge         1                             0    conda-forge
freetype                  2.10.4               h0708190_1    conda-forge
freexl                    1.0.6                h7f98852_0    conda-forge
gcc_impl_linux-64         11.2.0               h82a94d6_8    conda-forge
gcc_linux-64              11.2.0               h39a9532_1    conda-forge
geos                      3.9.1                h9c3ff4c_2    conda-forge
geotiff                   1.6.0                h4f31c25_6    conda-forge
gettext                   0.19.8.1          h73d1719_1006    conda-forge
gflags                    2.2.2             he1b5a44_1004    conda-forge
giflib                    5.2.1                h36c2ea0_2    conda-forge
glog                      0.5.0                h48cff8f_0    conda-forge
grpc-cpp                  1.40.0               h850795e_0    conda-forge
gxx_impl_linux-64         11.2.0               h82a94d6_8    conda-forge
gxx_linux-64              11.2.0               hacbe6df_1    conda-forge
hdf4                      4.2.15               h10796ff_3    conda-forge
hdf5                      1.12.1          nompi_h2750804_100    conda-forge
icu                       68.1                 h58526e2_0    conda-forge
jbig                      2.1               h7f98852_2003    conda-forge
jpeg                      9d                   h36c2ea0_0    conda-forge
json-c                    0.15                 h98cffda_0    conda-forge
kealib                    1.4.14               h87e4c3c_3    conda-forge
kernel-headers_linux-64   2.6.32              he073ed8_14    conda-forge
krb5                      1.19.2               hcc1bbae_0    conda-forge
lcms2                     2.12                 hddcbb42_0    conda-forge
ld_impl_linux-64          2.36.1               hea4e1c9_2    conda-forge
lerc                      2.2.1                h9c3ff4c_0    conda-forge
libarchive                3.5.2                hccf745f_0    conda-forge
libblas                   3.9.0           11_linux64_openblas    conda-forge
libbrotlicommon           1.0.9                h7f98852_5    conda-forge
libbrotlidec              1.0.9                h7f98852_5    conda-forge
libbrotlienc              1.0.9                h7f98852_5    conda-forge
libcblas                  3.9.0           11_linux64_openblas    conda-forge
libclang-cpp              11.1.0          default_ha53f305_1    conda-forge
libclang-cpp11.1          11.1.0          default_ha53f305_1    conda-forge
libcurl                   7.79.0               h2574ce0_0    conda-forge
libdap4                   3.20.6               hd7c4107_2    conda-forge
libdeflate                1.7                  h7f98852_5    conda-forge
libedit                   3.1.20191231         he28a2e2_2    conda-forge
libev                     4.33                 h516909a_1    conda-forge
libevent                  2.1.10               hcdb4288_3    conda-forge
libffi                    3.4.2                h9c3ff4c_1    conda-forge
libgcc-devel_linux-64     11.2.0               h0952999_8    conda-forge
libgcc-ng                 11.2.0               h1d223b6_8    conda-forge
libgcrypt                 1.9.3                h7f98852_1    conda-forge
libgdal                   3.3.2                h399a533_1    conda-forge
libgfortran-ng            11.2.0               h69a702a_8    conda-forge
libgfortran5              11.2.0               h5c6108e_8    conda-forge
libglib                   2.68.4               h174f98d_1    conda-forge
libgomp                   11.2.0               h1d223b6_8    conda-forge
libgpg-error              1.42                 h9c3ff4c_0    conda-forge
libgsasl                  1.10.0               h5b4c23d_0    conda-forge
libiconv                  1.16                 h516909a_0    conda-forge
libkml                    1.3.0             h238a007_1014    conda-forge
liblapack                 3.9.0           11_linux64_openblas    conda-forge
libllvm11                 11.1.0               hf817b99_2    conda-forge
libnetcdf                 4.8.1           nompi_hb3fd0d9_101    conda-forge
libnghttp2                1.43.0               h812cca2_0    conda-forge
libntlm                   1.4               h7f98852_1002    conda-forge
libopenblas               0.3.17          pthreads_h8fe5266_1    conda-forge
libpng                    1.6.37               h21135ba_2    conda-forge
libpq                     13.3                 hd57d9b9_0    conda-forge
libprotobuf               3.16.0               h780b84a_0    conda-forge
librdkafka                1.6.1                hc49e61c_1    conda-forge
librttopo                 1.1.0                h1185371_6    conda-forge
libsanitizer              11.2.0               he4da1e4_8    conda-forge
libspatialite             5.0.1                h8694cbe_5    conda-forge
libssh2                   1.10.0               ha56f1ee_0    conda-forge
libstdcxx-devel_linux-64  11.2.0               h0952999_8    conda-forge
libstdcxx-ng              11.2.0               he4da1e4_8    conda-forge
libthrift                 0.14.2               he6d91bd_1    conda-forge
libtiff                   4.3.0                hf544144_1    conda-forge
libutf8proc               2.6.1                h7f98852_0    conda-forge
libuuid                   2.32.1            h7f98852_1000    conda-forge
libwebp-base              1.2.1                h7f98852_0    conda-forge
libxcb                    1.13              h7f98852_1003    conda-forge
libxml2                   2.9.12               h72842e0_0    conda-forge
libzip                    1.8.0                h4de3113_0    conda-forge
lz4-c                     1.9.3                h9c3ff4c_1    conda-forge
lzo                       2.10              h516909a_1000    conda-forge
ncurses                   6.2                  h58526e2_4    conda-forge
numpy                     1.21.2           py39hdbf815f_0    conda-forge
omniscidb-common          5.7.0                h1234567_1    conda-forge
omniscidbe                5.7.0            h1234567_1_cpu    conda-forge
openjdk                   8.0.302              h7f98852_0    conda-forge
openjpeg                  2.4.0                hb52868f_1    conda-forge
openssl                   1.1.1l               h7f98852_0    conda-forge
orc                       1.6.10               h58a87f1_0    conda-forge
parquet-cpp               1.5.1                         2    conda-forge
pcre                      8.45                 h9c3ff4c_0    conda-forge
pip                       21.2.4             pyhd8ed1ab_0    conda-forge
pixman                    0.40.0               h36c2ea0_0    conda-forge
poppler                   21.09.0              h0e1ea10_1    conda-forge
poppler-data              0.4.11               hd8ed1ab_0    conda-forge
postgresql                13.3                 h2510834_0    conda-forge
proj                      8.0.1                h277dcde_0    conda-forge
pthread-stubs             0.4               h36c2ea0_1001    conda-forge
pyarrow                   3.0.0           py39h3ebc44c_24_cpu    conda-forge
pyomniscidbe              5.7.0           py39h1234567_1_cpu    conda-forge
python                    3.9.7           hb7a2778_1_cpython    conda-forge
python_abi                3.9                      2_cp39    conda-forge
re2                       2021.09.01           h9c3ff4c_0    conda-forge
readline                  8.1                  h46c0cb4_0    conda-forge
s2n                       1.0.10               h9b69904_0    conda-forge
setuptools                58.0.4           py39hf3d152e_0    conda-forge
snappy                    1.1.8                he1b5a44_3    conda-forge
sqlite                    3.36.0               h9cd32fc_1    conda-forge
sysroot_linux-64          2.12                he073ed8_14    conda-forge
tbb                       2021.3.0             h4bd325d_0    conda-forge
tbb4py                    2021.3.0         py39ha765c23_0    conda-forge
thrift-cpp                0.13.0               h5aa387f_5    conda-forge
tiledb                    2.3.4                he87e0bf_0    conda-forge
tk                        8.6.11               h27826a3_1    conda-forge
tzcode                    2021a                h7f98852_2    conda-forge
tzdata                    2021a                he74cb21_1    conda-forge
wheel                     0.37.0             pyhd8ed1ab_1    conda-forge
xerces-c                  3.2.3                h9d8b166_2    conda-forge
xorg-kbproto              1.0.7             h7f98852_1002    conda-forge
xorg-libice               1.0.10               h7f98852_0    conda-forge
xorg-libsm                1.2.3             hd9c2040_1000    conda-forge
xorg-libx11               1.7.2                h7f98852_0    conda-forge
xorg-libxau               1.0.9                h7f98852_0    conda-forge
xorg-libxdmcp             1.1.3                h7f98852_0    conda-forge
xorg-libxext              1.3.4                h7f98852_1    conda-forge
xorg-libxrender           0.9.10            h7f98852_1003    conda-forge
xorg-renderproto          0.11.1            h7f98852_1002    conda-forge
xorg-xextproto            7.3.0             h7f98852_1002    conda-forge
xorg-xproto               7.0.31            h7f98852_1007    conda-forge
xz                        5.2.5                h516909a_1    conda-forge
zlib                      1.2.11            h516909a_1010    conda-forge
zstd                      1.5.0                ha95c52a_0    conda-forge
```
</details>

<br/>
Details about  <code>conda</code> and system ( <code>conda info</code> ):
<details>

```
$ conda info

     active environment : test
    active env location : /home/user/miniconda3/envs/test
            shell level : 3
       user config file : /home/user/.condarc
 populated config files :
          conda version : 4.10.3
    conda-build version : not installed
         python version : 3.9.5.final.0
       virtual packages : __linux=4.4.0=0
                          __glibc=2.31=0
                          __unix=0=0
                          __archspec=1=x86_64
       base environment : /home/user/miniconda3  (writable)
      conda av data dir : /home/user/miniconda3/etc/conda
  conda av metadata url : None
           channel URLs : https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /home/user/miniconda3/pkgs
                          /home/user/.conda/pkgs
       envs directories : /home/user/miniconda3/envs
                          /home/user/.conda/envs
               platform : linux-64
             user-agent : conda/4.10.3 requests/2.25.1 CPython/3.9.5 Linux/4.4.0-18362-Microsoft ubuntu/20.04.2 glibc/2.31
                UID:GID : 1000:1000
             netrc file : None
           offline mode : False
```
</details>

cc @pearu

**Comment by tomashek** [comment]
Note that part of the problem is that the build requirements in this recipe require arrow 3.0 and thrift-cpp 0.13.
That thrift version predates the thrift recipe change that split libthrift into a separate package, so there are thrift-cpp packages available that DO NOT run-constrain the libthrift version (because libthrift did not exist when those were built).

Also, restricting arrow-cpp to version 3.0 and thrift 0.13 means only 6+ months old arrow-cpp packages can be used, and those have other library version problems (such as, they all depend on glog 0.4.* while omnisci requires 0.5.*).

Right now, the solver finds thrift-cpp 0.13 to satisfy omnisci, and libthrift 0.14 to satisfy arrow-cpp, and installs both.

Is it possible to build omnisci with thrift 0.14? Or if 0.13 is needed, can the thrift-cpp dependency be changed to libthrift? Or require both libthrift and thrift-cpp 0.13? If that constraint was added at build time, then the environment would have a consistent set of other libraries that could be used at runtime.

**Comment by pearu** [comment]
The latest omniscidb release 5.7.1 is thrift 0.14 ready and builds find. However, pyomniscidb requires thrift package from https://github.com/conda-forge/thrift-feedstock that provides only thrift 0.13. Hence, the issue is blocked until thrift 0.14 conda package will be available.

**Comment by vnlitvinov** [comment]
We have made a successful PR https://github.com/conda-forge/thrift-feedstock/pull/22, so `thrift-0.14` is available in conda-forge (or will be very soon when their CDN finishes propagating).

Can we get `thrift`-fixed build then? 😃

**Comment by pearu** [comment]
Thanks, @vnlitvinov, for pushing thrift-0.14 into conda-forge.

I have created https://github.com/conda-forge/omniscidb-feedstock/pull/54
(Btw, the pyomniscidb part will still use thrift-0.13 because upstream needs to be updated but that should not affect pyomniscidbe nor prevent resolving this issue).

**Comment by pearu** [comment]
Fixed via #54 :
```
$ conda create -n test-omniscidbe -c conda-forge pyomniscidbe
$ conda activate test-omniscidbe
$ conda list | grep thrift
libthrift                 0.14.2               he6d91bd_1    conda-forge
```
